numba/tests/test_tuples.py

```diff
@@ -609,6 +609,178 @@ def test_index(self):
         self.assertEqual(msg, str(raises.exception))
 
 
+class TestIntrinsics(TestCase):
+    def test_make_tuple_types(self):
+        from numba.targets.tupleobj import make_tuple
+    
+        @njit
+        def foo():
+            a = make_tuple(3, list(range(3)))
+            b = make_tuple(3, range(3))
+            c = make_tuple(3, zip(range(0, 3), range(3, 6)))
+            it = iter(range(3))
+            d = make_tuple(3, it)
+            arr = np.array(list(range(3)))
+            e = make_tuple(3, arr)
```

Comment:
might be worth adding a simple list comprehension, a typed-list and typed-dict to this.

Reply:
@esc Good point. I will do that. I am still a bit fuzzy though on typed lists and reflected lists: Are not all lists created in nopython mode typed? If yes, this would already be a typed-list.

Reply:
There are two types of list currently available in Numba, the so-called *reflected-list* and the *typed-list*.  You can read more about them and their differences here: http://numba.pydata.org/numba-doc/latest/reference/pysupported.html#list - let me know if the documentation leaves any unanswered questions.

Reply:
@esc Thanks. I thought this was just semantics to help the transition from python space. I did not realise they are internally different implementations.

Good call testing the overlooked dict types in particular: they were failing...